odxtools/encodestate.py

```diff
@@ -86,66 +87,145 @@ def emplace_atomic_value(
+            # note that we do not ensure that BCD-encoded byte fields
+            # only represent "legal" values
+            raw_value = bytes(internal_value)
 
-            raw_value = internal_value.encode("utf-8")
+        # ... string types, ...
+        elif base_data_type in (DataType.A_UTF8STRING, DataType.A_ASCIISTRING,
+                                DataType.A_UNICODE2STRING):
+            if not isinstance(internal_value, str):
+                odxraise(f"The internal value {internal_value!r} is not a string", EncodeError)
+
+            # note that the spec disallows certain combinations of
+            # base_data_type and encoding (e.g., A_ASCIISTRING encoded
+            # using UTF-8). Since in python3 strings are always
+            # capable of the full unicode character set, odxtools
+            # ignores these restrictions...
+
+            if base_type_encoding == Encoding.UTF8 or (base_data_type == DataType.A_UTF8STRING and
+                                                       base_type_encoding is None):
+                raw_value = internal_value.encode("utf-8")
+            elif base_type_encoding == Encoding.UCS2 or (base_data_type == DataType.A_UNICODE2STRING
+                                                         and base_type_encoding is None):
+                text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                raw_value = internal_value.encode(text_encoding)
+            elif base_type_encoding == Encoding.ISO_8859_1 or (
+                    base_data_type == DataType.A_ASCIISTRING and base_type_encoding is None):
+                raw_value = internal_value.encode("iso-8859-1")
+            elif base_type_encoding == Encoding.ISO_8859_2:
+                raw_value = internal_value.encode("iso-8859-2")
+            elif base_type_encoding == Encoding.WINDOWS_1252:
+                raw_value = internal_value.encode("cp1252")
+            else:
+                odxraise(f"Specified illegal encoding {base_type_encoding} for string object")
+                raw_value = internal_value.encode("iso-8859-1")
 
             if 8 * len(raw_value) > bit_length:
-                raise EncodeError(f"The string {repr(internal_value)} is too large."
-                                  f" The maximum number of bytes is {bit_length//8}.")
+                odxraise(
+                    f"The internal representation {raw_value!r} is too large "
+                    f"to be encoded. The maximum number of bytes is {(bit_length + 7)//8}.",
+                    EncodeError)
 
-        elif base_data_type == DataType.A_UNICODE2STRING:
-            if not isinstance(internal_value, str):
-                odxraise()
+        # ... signed integers, ...
+        elif base_data_type == DataType.A_INT32:
+            if not isinstance(internal_value, int):
+                odxraise(
+                    f"Internal value must be of integer type, not {type(internal_value).__name__}",
```

Comment:
How would this happen?

Reply:
for `.encode()`, `internal_value` is usually specified by the user?!

Reply:
User specify physical value, not internal one

Reply:
(see my comment below)